### PR TITLE
[GH-606][FEATURE] Add config mode for ModeWithIRSA to support using d…

### PR DESCRIPTION
**Pull request by stewartmoreland**
[GH-606][FEATURE] Add config mode for ModeWithIRSA to support using default credential provider with RUN_WITH_IRSA environment variable as True

# Description of the issue
Config mode OnPrem injects profile & file based credential configs to `amazon-cloudwatch-agent.toml`, thereby breaking support for using the default credential provider chain when running in EKS Anywhere or other k8s on-premises environments with IRSA.

Linked Issue: [GH-606](https://github.com/aws/amazon-cloudwatch-agent/issues/606)

# Description of changes
Change adds a new config mode (`ModeWithIRSA`) to prevent from adding profile & file based credentials and allowing the SDK to assume the default credential provider chain.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Unit tests have been executed and passed.

Agent has been deployed as DaemonSet to EKS Anywhere and validated functionality with `AssumeRoleWithWebIdentity` / IRSA.

### CODE REVIEW

File: translator/util/sdkutil.go [hunk 1]
```diff
@@ -38,6 +39,11 @@ func DetectAgentMode(configuredMode string) string {
 		return config.ModeEC2
 	}
 
+	if runWithIrsa == config.RUN_WITH_IRSA_TRUE {
```
Comment (khanhntd) [inline translator/util/sdkutil.go:42]: Would there be any reason why this won't be the last check? (e.g after `if DefaultECSRegion() != ""{)`)

**Review by khanhntd — COMMENTED** [review]
Would need to think more deeply about the implementation though ! However, thanks for the contribution and we are very appreciated for it.

File: translator/context/context.go [hunk 2]
```diff
@@ -114,8 +114,11 @@ func (ctx *Context) SetMode(mode string) {
 		ctx.mode = config.ModeOnPrem
 	case config.ModeOnPremise:
 		ctx.mode = config.ModeOnPremise
+		ctx.mode = config.ModeOnPrem
```
[the remark this answers is not in the document]
Comment (stewartmoreland) [reply translator/context/context.go]: Correct. This was unintentional. Updated latest commit removing this line. Thank you!

File: translator/util/sdkutil.go (hunk 1 above)
Comment (khanhntd) [reply translator/util/sdkutil.go:42]: No. Why would this check get prioritized  over all other checks and not as a last check before return its as onPrem

Comment (khanhntd) [reply translator/util/sdkutil.go:42]: e.g Why not putting it here
```
if DefaultEC2Region() != "" {
		fmt.Println("I! Detected the instance is EC2")
		return config.ModeEC2
	}
	if DefaultECSRegion() != "" {
		fmt.Println("I! Detected the instance is ECS")
		return config.ModeEC2
	}
        if runWithIrsa == config.RUN_WITH_IRSA_TRUE {
               return config.ModeWithIRSA
        }
	fmt.Println("I! Detected the instance is OnPremise")
	return config.ModeOnPrem
```
but the reverse?

File: translator/config/defaultConfig.go [hunk 3]
```diff
@@ -195,6 +195,8 @@ func DefaultJsonConfig(os string, mode string) string {
 	default:
 		if mode == ModeEC2 {
 			return defaultLinuxEC2Config
+		} else if mode == ModeWithIRSA {
+			return defaultLinuxOnPremConfig
```
Comment (SaxyPandaBear) [inline translator/config/defaultConfig.go:199]: I think that this case is captured in the `else` block already

File: translator/util/sdkutil.go (hunk 1 above)
Comment (SaxyPandaBear) [reply translator/util/sdkutil.go:42]: Because the RUN_WITH_IRSA is an override flag. It should come before we try to ping IMDS and ECS

**Review by SaxyPandaBear — APPROVED** [review]

File: translator/util/sdkutil.go (hunk 1 above)
Comment (khanhntd) [reply translator/util/sdkutil.go:42]: Sounds fair to me !

**Review by khanhntd — APPROVED** [review]
IMO, using flag would not be recommended  since its a hack way and needs to be known for every customers. Therefore, it's not an automated mechanism. However, I'm good with this shorted term solution since its benefit for all EKS customers running `onPremise`. Since its passing the sanity and its easy to roll back, I'm good with it.

**Review by khanhntd — APPROVED** [review]
LGTM. Thanks for the contribution and great deep dive on it !

**Review by SaxyPandaBear — APPROVED** [review]
LGTM

**Pull request merged**